### PR TITLE
Add nonfungibles::Create Trait & Implement for Uniques

### LINKED ISSUE #9419: Why Doesn't nonfungibles::Mutate Specify a create_class Function?
I'm having a little trouble understanding how to use the Uniques pallet, and it seems like one of the most straightforward ways to resolve my confusion would be to add a `create_class` function to the `nonfungibles::Mutate` trait. Otherwise, it seems entirely necessary for class IDs to be tracked off-chain, and that seems to be of limited utility. Why doesn't `nonfungibles::Mutate` specify a function to create an asset class? Alternatively/additionally, what is the intended use of the Uniques pallet with respect to making it easy for users to create asset classes with unique IDs?

### PR DESCRIPTION
**Pull request by danforbes**
Closes #9419

Please let me know if this has any testing, benchmarking, or documentation implications that I missed.

### CODE REVIEW

File: frame/support/src/traits/tokens/nonfungibles.rs [hunk 1]
```diff
@@ -117,6 +117,18 @@ pub trait InspectEnumerable<AccountId>: Inspect<AccountId> {
...
+/// Trait for providing the ability to create classes of nonfungible assets.
+pub trait Create<AccountId>: Inspect<AccountId> {
+	/// Create a `class` of nonfungible assets to be owned by `who` and managed by `admin`.
+	fn create_class(
+		_class: &Self::ClassId,
+		_who: &AccountId,
+		_admin: &AccountId,
+	) -> DispatchResult {
+		Err(TokenError::Unsupported.into())
+	}
```
Comment (bkchr) [inline frame/support/src/traits/tokens/nonfungibles.rs:129]:
```suggestion
	fn create_class(
		class: &Self::ClassId,
		who: &AccountId,
		admin: &AccountId,
	) -> DispatchResult;
```

File: frame/uniques/src/functions.rs [hunk 2]
```diff
@@ -48,6 +48,39 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		Ok(())
 	}
 
+	pub(super) fn do_create_class(
+		class: T::ClassId,
+		owner: T::AccountId,
+		admin: T::AccountId,
+		deposit: Option<DepositBalanceOf<T, I>>,
+		free_holding: Option<bool>,
```
Comment (apopiak) [inline frame/uniques/src/functions.rs:56]: I would just pass these as values directly.

**Review by apopiak — COMMENTED** [review]
Looks fine to me, would remove the option wrapping of the parameters.

**Review by apopiak — APPROVED** [review]
LGTM

File: frame/uniques/src/lib.rs [hunk 3]
```diff
@@ -366,25 +352,14 @@ pub mod pallet {
...
-					instances: 0,
-					instance_metadatas: 0,
-					attributes: 0,
-					is_frozen: false,
-				},
-			);
-			Self::deposit_event(Event::ForceCreated(class, owner));
-			Ok(())
+				owner.clone(),
+				owner.clone(),
```
Comment (shawntabrizi) [inline frame/uniques/src/lib.rs:358]: clone shouldnt be needed here i think

File: frame/uniques/src/lib.rs [hunk 4]
```diff
@@ -316,28 +316,14 @@ pub mod pallet {
...
-					instances: 0,
-					instance_metadatas: 0,
-					attributes: 0,
-					is_frozen: false,
-				},
-			);
-			Self::deposit_event(Event::Created(class, owner, admin));
-			Ok(())
+				owner.clone(),
+				admin.clone(),
```
Comment (shawntabrizi) [inline frame/uniques/src/lib.rs:322]: same here

File: frame/uniques/src/impl_nonfungibles.rs [hunk 5]
```diff
@@ -85,6 +88,24 @@ impl<T: Config<I>, I: 'static> Inspect<<T as SystemConfig>::AccountId> for Palle
...
+	/// Create a `class` of nonfungible assets to be owned by `who` and managed by `admin`.
+	fn create_class(
+		class: &Self::ClassId,
+		who: &T::AccountId,
+		admin: &T::AccountId,
+	) -> DispatchResult {
+		Self::do_create_class(
+			class.clone(),
+			who.clone(),
+			admin.clone(),
```
Comment (shawntabrizi) [inline frame/uniques/src/impl_nonfungibles.rs:101]: and here

**Review by shawntabrizi — APPROVED** [review]
looks okay besides some maybe un-needed clones. please fix up before merge

**Comment by danforbes** [comment]
@shawntabrizi - if I remove the clones, the build fails.

File: frame/uniques/src/lib.rs (hunk 3 above)
Comment (apopiak) [reply frame/uniques/src/lib.rs:358]: Those clones look fine to me. `do_create_class` takes params by value, the same as the other functions in `functions.rs`.

**Comment by apopiak** [comment]
The clones look fine to me.

**Comment by apopiak** [comment]
bot merge

**Pull request merged**